langchain/src/callbacks/handlers/tracers.ts

```diff
@@ -371,11 +413,102 @@ export class LangChainTracer extends BaseTracer {
+    let runResult: LLMRun | ChainRun | ToolRun;
+    if (run.run_type === "llm") {
+      const llmRun: LLMRun = {
+        uuid: run.id,
+        start_time: run.start_time,
+        end_time: run.end_time,
+        execution_order: run.execution_order,
+        child_execution_order: run.child_execution_order,
+        serialized,
+        type: run.run_type,
+        session_id: session.id,
+        prompts: run.inputs.prompts,
+        response: run.outputs as LLMResult,
+      };
+      runResult = llmRun;
+    } else if (run.run_type === "chain") {
+      const child_runs = await Promise.all(
+        run.child_runs.map((child_run) => this.convertV2RunToRun(child_run))
+      );
+      const chainRun: ChainRun = {
+        uuid: run.id,
+        start_time: run.start_time,
+        end_time: run.end_time,
+        execution_order: run.execution_order,
+        child_execution_order: run.child_execution_order,
+        serialized,
+        type: run.run_type,
+        session_id: session.id,
+        inputs: run.inputs,
+        outputs: run.outputs,
+        child_llm_runs: child_runs.filter(
+          (child_run) => child_run.type === "llm"
+        ) as LLMRun[],
+        child_chain_runs: child_runs.filter(
+          (child_run) => child_run.type === "chain"
+        ) as ChainRun[],
+        child_tool_runs: child_runs.filter(
+          (child_run) => child_run.type === "tool"
+        ) as ToolRun[],
+      };
+
+      runResult = chainRun;
+    } else if (run.run_type === "tool") {
+      const child_runs = await Promise.all(
+        run.child_runs.map((child_run) => this.convertV2RunToRun(child_run))
+      );
+      const toolRun: ToolRun = {
+        uuid: run.id,
+        start_time: run.start_time,
+        end_time: run.end_time,
+        execution_order: run.execution_order,
+        child_execution_order: run.child_execution_order,
+        serialized,
+        type: run.run_type,
+        session_id: session.id,
+        tool_input: run.inputs.input,
+        output: run.outputs.output,
+        action: JSON.stringify(serialized),
+        child_llm_runs: child_runs.filter(
+          (child_run) => child_run.type === "llm"
+        ) as LLMRun[],
+        child_chain_runs: child_runs.filter(
+          (child_run) => child_run.type === "chain"
+        ) as ChainRun[],
+        child_tool_runs: child_runs.filter(
+          (child_run) => child_run.type === "tool"
+        ) as ToolRun[],
+      };
+
+      runResult = toolRun;
+    } else {
+      throw new Error(`Unknown run type: ${run.run_type}`);
+    }
+    return runResult;
+  }
+
+  protected async persistRun(
+    run: RunV2 | LLMRun | ChainRun | ToolRun
+  ): Promise<void> {
     let endpoint;
-    if (run.type === "llm") {
+    let v1Run: LLMRun | ChainRun | ToolRun;
+    if ((run as RunV2).run_type !== undefined) {
+      v1Run = await this.convertV2RunToRun(run as RunV2);
+    } else {
+      v1Run = run as LLMRun | ChainRun | ToolRun;
```

Comment:
How would we get here?

Reply:
It should never get here. I could actually probably update the `persistRun` argument typing  - thanks for pointing that out